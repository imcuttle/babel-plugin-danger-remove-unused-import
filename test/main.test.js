/**
 * @file main
 * @author Cuttle Cong
 * @date 2018/11/5
 * @description
 */
const nps = require('path')
const fs = require('fs')
const { transform } = require('babel-core')

function fixture(name) {
  return nps.join(__dirname, 'fixture', name)
}

function read(name) {
  return fs.readFileSync(fixture(name), { encoding: 'utf8' })
}

function transformTest(name, opts, { plugins = [], ...rest } = {}) {
  return transform(read(name), {
    plugins: [[require('../src'), opts]].concat(plugins),
    babelrc: false,
    ...rest
  })
}

describe('main', function() {
  describe('unused', () => {
    it('JSX', () => {
      expect(
        transformTest(
          'unused/JSX.js',
          {},
          { presets: [require.resolve('babel-preset-react')] }
        ).code
      ).toMatchSnapshot()
    })

    it('VariableDeclarator', () => {
      expect(
        transformTest('unused/VariableDeclarator.js', {}).code
      ).toMatchSnapshot()
    })

    it('LabeledStatement', () => {
      expect(transformTest('unused/LabeledStatement.js').code).toMatchSnapshot()
    })

    it('ObjectProperty', () => {
      expect(
        transformTest(
          'unused/ObjectProperty.js',
          {},
          { presets: ['babel-preset-stage-0'] }
        ).code
      ).toMatchSnapshot()
    })

    it('MemberExpression', () => {
      expect(transformTest('unused/MemberExpression.js').code)
        .toMatchInlineSnapshot(`
"

const ref = {};
const x = ref.Tab;"
`)
    })

    it('MemberExpression-2', () => {
      expect(transformTest('unused/MemberExpression-2.js').code)
        .toMatchInlineSnapshot(`
"

const ref = {};
const x = ref['Tab'];"
`)
    })

    it('class', function() {
      expect(transformTest('unused/class.js').code).toMatchInlineSnapshot(`
"

class A {
  Tab() {}

  static Tab() {}
}"
`)
    })
  })

  describe('used', () => {
    it('JSX', function() {
      expect(transformTest('used/JSX.js', {}, { presets: ['react'] }).code)
        .toMatchInlineSnapshot(`
"import Tab from 'tab';

const comp = React.createElement(Tab, null);"
`)
    })

    it('ObjectProperty', function() {
      expect(
        transformTest(
          'used/ObjectProperty.js',
          {},
          { presets: ['babel-preset-stage-0'] }
        ).code
      ).toMatchSnapshot()
    })

    it('class', function() {
      expect(
        transformTest(
          'used/class.js',
          {},
          { presets: ['babel-preset-stage-0'] }
        ).code
      ).toMatchInlineSnapshot(`
"import Tab from 'tab';

export class A {
  [Tab]() {}
}"
`)
    })

    it('class-2', function() {
      expect(transformTest('used/class-2.js', {}, { presets: ['react'] }).code)
        .toMatchInlineSnapshot(`
"import Tab from 'tab';

export class A {
  static [Tab]() {}
}"
`)
    })
  })

  describe('options', () => {
    it('options `ignore = [react]`', function() {
      expect(
        transformTest(
          'options/ignore.js',
          { ignore: ['react'] },
          { presets: ['react'] }
        ).code
      ).toMatchInlineSnapshot(`
"import * as React from 'react';

export default (() => React.createElement(
  'div',
  null,
  'hah'
));"
`)
    })

    it('options `ignore = []`', function() {
      expect(
        transformTest(
          'options/ignore.js',
          { ignore: [] },
          { presets: ['react'] }
        ).code
      ).toMatchInlineSnapshot(`
"

export default (() => React.createElement(
  'div',
  null,
  'hah'
));"
`)
    })
  })
})
