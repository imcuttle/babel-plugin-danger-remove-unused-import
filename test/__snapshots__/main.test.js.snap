// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main unused JSX 1`] = `
"

const comp = React.createElement(a.Tab, null);"
`;

exports[`main unused LabeledStatement 1`] = `
"

{
  Tab: 'abc';
}"
`;

exports[`main unused ObjectProperty 1`] = `
"var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

const x = _extends({ Tab: 'abc' }, { a: '2' });"
`;

exports[`main unused VariableDeclarator 1`] = `
"

(function () {
  const Tab = '123';
})();"
`;

exports[`main used ObjectProperty 1`] = `
"var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

import Tab from 'tab';

export const x = _extends({ [Tab]: 'abc' }, { a: '2' });"
`;
